[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops

anshul-si opened the pull request:
**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170632
* #170040
* __->__ #170035
* #170030



cc @wanchaol @tianyu-l @wz337 @XilunWu @d4l3k @pragupta @SherlockNoMad @ppwwyyxx

anshul-si committed:
[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -534,14 +535,23 @@ def common_pointwise_strategy(
-                    safe_avoid_redistribution = True
+                elif isinstance(placement, Partial):
+                    if op not in redistribute_partial_ops or not scalar_arg:
+                        safe_avoid_redistribution = True
```
building on my comment in the previous PR, maybe a more understandable way to write this one is

```
if linearity >= 0 and partial_supports_linearity and not special_case_add_scalar and not special_case_normpartial:
```

Alternatively, if we break the subclass relationship we end up with 2 top level elif cases

elif isinstance(p, Partial)
  # no normpartial special case in here
elif isinstance(p, _NormPartial)
  # no add special case in here

anshul-si replied:
I like the idea of breaking the sub class relationship.

wconstab commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -631,10 +641,25 @@ def common_pointwise_strategy(
-    )
+    if op in norm_partial_avoidable_redistribute_ops:
+        register_op_strategy(
+            op, schema_info=RuntimeSchemaInfo(1, static_kwargkey=["out"])
```
remind me what the meaning of this `1` is for caching? and why it relates to NormPartial

anshul-si commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -12,6 +12,7 @@
     StrategyType,
     TupleStrategy,
 )
+from torch.distributed.tensor._ops._math_ops import _NormPartial
```
I think that we should make it so that atleast is_partial returns False for _NormPartial. Correct me if I'm wrong, but I think this would mean we would need a set of new sharding propagation rules for _NormPartial. This makes sense to me anyways since I know of atleast 2-3 rules that don't apply to NormPartial in different files (i.e pointwise_ops and math_ops).

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

wconstab commented:
i guess this CI failure is unrelated, but you should rebase to make sure.

> ____________________ InplacePaddingTest.test_linear_and_cel ____________________
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/inductor/test_inplace_padding.py", line 247, in test_linear_and_cel
    self.assertEqual(num_inplace_padding(), int(inductor_config.inplace_padding))
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/test_case.py", line 112, in assertEqual
    return super().assertEqual(x, y, *args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 4284, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Scalars are not equal!

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]

anshul-si committed:
Update on "[dtensor][partial] redistributes _NormPartial to replicate for necessary pointwise_ops"

**Summary:** While linearities are correct for regular Partial tensors, they don't apply to NormPartial tensors. We fix this by redistributing for all ops necessary when the placement is NormPartial. The math showing that mul and div scalar ops don't need redistribution when the scalar value is non negative is shown below. The iterative process to arriving at this PR can be viewed in https://github.com/pytorch/pytorch/pull/167813. 

<img width="449" height="807" alt="image" src="https://github.com/user-attachments/assets/5e42d65d-d05f-43eb-9d69-9e663fb1f1eb" />

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_mul_div_scalar_norm_partial
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_scalar_norm_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad ppwwyyxx

[ghstack-poisoned]
